kernel/arch/riscv/early_fdt.c

```diff
@@ -0,0 +1,395 @@
+};
+
+struct simplefb_format {
+   const char *name;
+   u8 bpp;  /* bits per pixel */
+   struct simplefb_bitfield red;
+   struct simplefb_bitfield green;
+   struct simplefb_bitfield blue;
+   struct simplefb_bitfield alpha;
+};
+
+void *fdt_blob;   //save virtual address of fdt
+
+static ulong initrd_paddr;
+static ulong initrd_size;
+
+static multiboot_info_t *mbi;
+static multiboot_module_t *mod;
+static multiboot_memory_map_t *mmmap;
+
+static int mmmap_count = 0;
+static char *cmdline_buf;
+
+static struct simplefb_format simplefb_formats[] = {
+   {"r5g6b5", 16, {11, 5}, {5, 6}, {0, 5}, {0, 0}},
+   {"r5g5b5a1", 16, {11, 5}, {6, 5}, {1, 5}, {0, 1}},
+   {"x1r5g5b5", 16, {10, 5}, {5, 5}, {0, 5}, {0, 0}},
+   {"a1r5g5b5", 16, {10, 5}, {5, 5}, {0, 5}, {15, 1}},
+   {"r8g8b8", 24, {16, 8}, {8, 8}, {0, 8}, {0, 0}},
+   {"x8r8g8b8", 32, {16, 8}, {8, 8}, {0, 8}, {0, 0}},
+   {"a8r8g8b8", 32, {16, 8}, {8, 8}, {0, 8}, {24, 8}},
+   {"x8b8g8r8", 32, {0, 8}, {8, 8}, {16, 8}, {0, 0}},
+   {"a8b8g8r8", 32, {0, 8}, {8, 8}, {16, 8}, {24, 8}},
+   {"x2r10g10b10", 32, {20, 10}, {10, 10}, {0, 10}, {0, 0}},
+   {"a2r10g10b10", 32, {20, 10}, {10, 10}, {0, 10}, {30, 2}},
+};
+
+void alloc_mbi(void)
+{
+   void *mbi_memtop;
+
+   /*
+    * Place multiboot information at the top of
+    * 8 MB identity mapping space.
+    */
+   mbi_memtop = (void *)(LIN_VA_TO_PA(BASE_VA) + EARLY_MAP_SIZE);
+
+   mbi = (multiboot_info_t *)((ulong)mbi_memtop - sizeof(*mbi));
+   bzero(mbi, sizeof(*mbi));
+
+   mod = (multiboot_module_t *)((ulong)mbi - sizeof(*mod));
+   bzero(mod, sizeof(*mod));
+
+   cmdline_buf = (char *)mod - CMDLINE_BUF_SZ;
+   bzero(cmdline_buf, CMDLINE_BUF_SZ);
+}
+
+static void
+add_multiboot_mmap(u64 addr, u64 size, u32 type)
+{
+   /* Allocate a new multiboot_memory_map */
+   mmmap = (void *)(cmdline_buf - \
+                     (++mmmap_count) * sizeof(multiboot_memory_map_t));
+
+   bzero(mmmap, sizeof(multiboot_memory_map_t));
+
+   if (addr < mbi->mem_lower * KB)
+      mbi->mem_lower = (u32)(addr / KB);
+
+   if (addr + size > mbi->mem_upper * KB)
+      mbi->mem_upper = (u32)((addr + size) / KB);
+
+   *mmmap = (multiboot_memory_map_t) {
+      .size = sizeof(multiboot_memory_map_t) - sizeof(u32),
+      .addr = addr,
+      .len = size,
+      .type = type,
+   };
+}
+
+void setup_multiboot_info(ulong ramdisk_paddr, ulong ramdisk_size)
+{
+   mbi->flags |= MULTIBOOT_INFO_MEMORY;
+
+   if (cmdline_buf[0]) {
+      mbi->flags |= MULTIBOOT_INFO_CMDLINE;
+      mbi->cmdline = (ulong)cmdline_buf;
+   }
+
+   mbi->flags |= MULTIBOOT_INFO_MODS;
+   mbi->mods_addr = (ulong)mod;
+   mbi->mods_count = 1;
+   mod->mod_start = ramdisk_paddr;
+   mod->mod_end = mod->mod_start + ramdisk_size;
+
+   mbi->flags |= MULTIBOOT_INFO_MEM_MAP;
+   mbi->mmap_addr = (ulong)mmmap;
+   mbi->mmap_length = mmmap_count * sizeof(multiboot_memory_map_t);
+}
+
+static int
+fdt_get_node_linux_usable_memory(void *fdt, int node, int index,
```

Comment:
Style: do you mind aligning the beginning of the next line of parameters with the first char after '(' ? 
E.g.
```
foo(a, b, c,
    d, e, f)
```

Alternatively, even better, you can put one parameter per line, plus the alignment as described. That is very convenient for adding comments next to the parameters.

Reply:
fixed!